zne/extrapolation/extrapolator.py

```diff
@@ -15,96 +15,123 @@
-            metadata = {"variance": variance, **metadata}
-        return prediction, metadata
-
-    def extrapolate_zero(self, data: Sequence[Sequence[float]]) -> tuple[float, Metadata]:
-        """Same as `infer` but fixing `target = 0`."""
-        return self.infer(0, data)
+        # Single validation
+        x_data = self._validate_data(x_data)
+        y_data = self._validate_data(y_data)
+        sigma_x = self._validate_sigma(sigma_x, default_size=x_data.size)
+        sigma_y = self._validate_sigma(sigma_y, default_size=y_data.size)
+        # Cross-validation
+        num_points = len(set(x_data))
+        if num_points < self.min_points:
+            raise ValueError(
+                f"Insufficient number of distinct data points provided ({num_points}), "
+                f"at least {self.min_points} needed."
+            )
+        sizes = (x_data.size, y_data.size, sigma_x.size, sigma_y.size)
+        if any(s != x_data.size for s in sizes):
+            raise ValueError(
+                "Invalid data, all inputs should be of the same size: "
+                f"{x_data.size = }, {y_data.size = }, {sigma_x.size = }, {sigma_y.size = }."
+            )
+        return self._extrapolate_zero(x_data, y_data, sigma_x, sigma_y)
 
     ################################################################################
     ## VALIDATION
     ################################################################################
-    def _validate_data(self, data: Sequence[Sequence[float]]) -> tuple[RegressionDatum, ...]:
-        """Validates data points for the regression model.
+    def _validate_data(
+        self, data: Sequence[float] | "ndarray[Any, dtype[float_]]"
+    ) -> "ndarray[Any, dtype[float_]]":
+        """Validates data for the regression model.
 
         Args:
-            data: A sequence of data points to fit the regression model from, each of which
-                consisting on the value of the regressor, measured mean, and estimated variance.
+            data: A sequence of values for the regression model.
 
         Returns:
             Validated data normalized.
-
-        Raises:
-            TypeError: If provided data is not a valid sequence of data points.
-            ValueError: If the number of data points is smaller than the minimum required
-                number of data points for the given Estrapolator.
         """
-        if not isinstance(data, Sequence) or not all(
-            isinstance(d, Sequence) and len(d) == 3 and all(isinstance(v, (float, int)) for v in d)
-            for d in data
-        ):
-            raise TypeError(
-                "Invalid data provided, expeceted Sequence[Sequence[float, float, float]]."
-            )
-        if len(data) < self.min_points:
+        try:
+            data = array(data).astype(float_, casting="same_kind")
+        except TypeError as exc:
+            raise TypeError(f"Invalid non-numeric (i.e. non-real) data array: {data}.") from exc
```

Comment:
Does this catch `np.nan`? I haven't checked whether there is a test for it yet, but `type(np.nan) == float`. Do you even want to catch NaN's?

Reply:
Good catch!

Reply:
New validation function along with the signature's types redefinition. This should be solved now, thanks!